Add support for ECR Lifecycle Policies to ecs_ecr

SpamapS opened the pull request:
##### SUMMARY

Add support for ECR Lifecycle Policies to ecs_ecr. Fixes #32003

##### ISSUE TYPE

    Feature Pull Request

##### COMPONENT NAME

ecs_ecr

##### ANSIBLE VERSION

ansible 2.4.0.0
  config file = /Users/dlee/.ansible.cfg
  configured module search path = [u'/Users/dlee/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.4.0.0/libexec/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]

##### ADDITIONAL INFORMATION

Integration tests provided, and they pass.

PLAY RECAP *********************************************************************
localhost                  : ok=55   changed=8    unreachable=0    failed=0

leedm777 committed:
Fix copy/pasta for ecs_ecr test names

leedm777 committed:
Add support for lifecycle policies to ecs_ecr

New feature for ecs_ecr to support [ECR Lifecycle Policies][].

Fixes #32003

 [ECR Lifecycle Policies]: https://docs.aws.amazon.com/AmazonECR/latest/userguide/LifecyclePolicies.html

leedm777 committed:
Improve error message for ecs_ecr parsing errors

Replaces the exception and stack trace with a description of what's
actually going wrong from a user perspective.

leedm777 committed:
Rename delete policy to purge policy

Marks the `delete_policy` parameter as deprecated, to be removed in
Ansible 2.6.

leedm777 committed:
Add version_added to purge_policy

leedm777 committed:
Remove changing results based on verbosity

What I really want is --diff support, and changing results based on
verbosity is abnormal.

s-hertel committed:
Ensure repository name is lowercase

s-hertel committed:
Fix deprecation cycle to 4 releases

s-hertel committed:
Use a YAML anchor for credentials

s-hertel committed:
Remove filters from assertions

s-hertel committed:
Add minimal permissions needed

SpamapS committed:
Updating version_added and deprecation cycle

The original PR sat while a few releases happened.

Maelstromeous commented:
Would be most appreciated if this PR could be picked back up :)

SpamapS committed:
Bumping version added and deprecation version

We missed the 2.8 release.

SpamapS commented:
Rebased! I've had this vendored locally since the initial rebase I did... would be great to rejoin the fold in some future Ansible release. :)

SpamapS committed:
Removing bare except:

This is not allowed and is generally bad practice.

SpamapS committed:
Fix lint errors

gundalow commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -43,11 +42,29 @@
         required: false
         default: false
         type: bool
+        version_added: '2.9'
```
```suggestion
        version_added: '2.10'
```

gundalow commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -43,11 +42,29 @@
+        description:
+            - JSON or dict that represents the new lifecycle policy
+        required: false
+        version_added: '2.9'
```
```suggestion
        version_added: '2.10'
```

gundalow commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -43,11 +42,29 @@
+            - if yes, remove the lifecycle policy from the repository
+        required: false
+        default: false
+        version_added: '2.9'
```
```suggestion
        version_added: '2.10'
```

gundalow commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -43,11 +42,29 @@
     delete_policy:
+        description:
+            - deprecated. Use I(purge_policy) instead. To be removed in
+              Ansible 2.13
```
```suggestion
              Ansible 2.14
```

SpamapS committed:
update ansible release metadata

SpamapS commented:
@gundalow thanks, committed your suggestions. :)

jillr commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -260,19 +341,34 @@ def sort_lists_of_strings(policy):
         state = params['state']
         policy_text = params['policy']
-        delete_policy = params['delete_policy']
+        purge_policy = params['purge_policy'] or params['delete_policy']
```
This didn't exist when this PR was first opened, but we now support alias deprecation. Could you please make `delete_policy` an alias for `purge_policy` instead of a seperate param, and add a deprecation like in https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/copy.py#L517:L518?  
Then a note in the module docs that `delete_policy` is deprecated.

SpamapS replied:
Done.

jillr commented on test/integration/targets/ecs_ecr/tasks/main.yml:
```diff
@@ -1,16 +1,22 @@
 ---
 - set_fact:
-    ecr_name: '{{ resource_prefix }}-ecr'
+    ecr_name: '{{ resource_prefix | lower }}-ecr'
```
This shouldn't be necessary as the resource_prefix is only ints and already lowered values. https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/_internal/cloud/__init__.py#L398

SpamapS replied:
I think you're right. Removed.

jillr commented on test/integration/targets/ecs_ecr/tasks/main.yml:
```diff
@@ -1,16 +1,22 @@
 
+    - name: set connection information for all tasks
+      set_fact:
+        aws_connection_info: &aws_connection_info
```
This is fine for this PR but we're trying to move the AWS tests to use module_defaults going forward.

SpamapS replied:
I don't understand that, and have not changed anything as a result.

jillr replied:
Just kind of an FYI that we're changing the way we do this, as we'd very much like to have your contributions going forward. Not asking for any changes in this case though cause I totally appreciate how long this PR has dragged on.

jillr commented on test/integration/targets/ecs_ecr/tasks/main.yml:
```diff
@@ -246,12 +229,168 @@
+      assert:
+        that:
+          - result is changed
+          - not result.created
```
Could you please add assertions about `result.lifecycle_policy` to the tests that act on livecycle policies?  Key exists and length of `rules` would be good, contents of the rule would be a bonus.

SpamapS replied:
Done, I think.

jillr commented on test/integration/targets/ecs_ecr/tasks/main.yml:
```diff
@@ -261,14 +400,11 @@
 
 
-    - name: When in check mode, deleting a policy that exists
+    - name: When in check mode, deleting a repository that exists
```
Thanks very much for catching and cleaning these up!

jillr reviewed (commented):
Hi @SpamapS, so sorry about the delay on this PR! We'd definitely like to get this merged in, appreciate your patience very much.  Just a few comments.

SpamapS committed:
Use the new alias deprecation scheme

This was added in the time the PR has been in development, so rework
things to use it.

SpamapS committed:
Add test coverage

This makes sure that lifecycle_policy is produced when passed in.

*Also a minor suggestion for simplification from PR.

SpamapS reviewed (commented):
Rebased *AGAIN*.

We're at > 1 year of rebasing and then not getting this in. It's quite a treadmill. If somebody can please prioritize landing this, I'm super happy to be able to contribute, but I'm also losing interest as we've moved to managing ecr policies in Terraform at this point.

jillr commented:
@SpamapS I'm looking info the test failures, I'll try to push an update to your branch if that works ok for you.

SpamapS commented:
I'll take a look at the test failures later today.

SpamapS commented:
But yes, I'll make sure you can push to my fork @jillr , first one to the fixes wins. ;)

jillr commented:
Looks like 62871 got partially lost in the rebase - I'll work on getting that added back in.

jillr committed:
Restore changes from 62871 lost in rebase

jillr commented:
I believe that should do it, sanity and integration tests are passing locally.  If CI passes and you don't have any objections to my changes there @SpamapS I'll give it a fresh set of eyes in the morning just to be safe but I think we should be about there.

jillr committed:
Add changelog

SpamapS committed:
Remove version_added for new purge_policy option

Per sanity test fail.

jillr reviewed (approved):
LGTM, thanks very much @SpamapS!
